### PR TITLE
Add support for ingesting data from Dataverse. Fixes #179

### LINKED ISSUE #179: Add support for registering data from various Dataverse installations
A few of us talked with @pdurbin at the 2018 workshop about adding support for Dataverse's API in WholeTale's Data Registration feature. This looks really easy. One note is that there are many Dataverse installations and no central location to query across all of them at once but they appear to all follow a common convention so I think we can easily support Harvard's installation but with only a bit more effort, any installation:

- WholeTale sees the DOI: 10.7910/DVN/TJCLKP
- And follows it to: https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/TJCLKP
- Then we can try the API: https://dataverse.harvard.edu/api/datasets/:persistentId?persistentId=doi:10.7910/DVN/TJCLKP

which gives a nicely-structured response body:

```json
{
  "status": "OK",
  "data": {
    "id": 3035124,
    "identifier": "DVN/TJCLKP",
    "persistentUrl": "https://doi.org/10.7910/DVN/TJCLKP",
...snip...
   "files": [
        {
          "label": "2017-07-31.tab",
...snip...
```

which has the necessary things: Dataset title, and for all files: filename, and size.

### PR DESCRIPTION
**Pull request by Xarthisius**
This PR introduces initial support for registering data from Dataverse using DOI resolver, e.g.:

- https://dx.doi.org/doi:10.7910/DVN/TJCLKP/BTRTVA (should register a single file)
- https://dx.doi.org/doi:10.7910/DVN/TJCLKP (should register a dataset)

A set of dataverse installation can be changed via `DATAVERSE_URL` config variable.

TODO:
- [x] add tests
- [x] allow to register data using internal ids
- [x] allow to point to a single dataverse instance via `DATAVERSE_URL`

### CODE REVIEW

File: server/constants.py [hunk 1]
```diff
@@ -24,12 +24,15 @@ class PluginSettings:
...
+    DATAVERSE_URL = 'wholetale.dataverse_url'
 
 
 class SettingDefault:
     defaults = {
         PluginSettings.INSTANCE_CAP: 2,
-        PluginSettings.DATAONE_URL: 'https://cn.dataone.org/cn/v2/node'
+        PluginSettings.DATAONE_URL: 'https://cn.dataone.org/cn/v2/node',
+        PluginSettings.DATAVERSE_URL:
+            'https://services.dataverse.harvard.edu/miniverse/map/installations-json'
```
Comment (pdurbin) [inline server/constants.py]: This "miniverse" URL may change some day but we'll try to let you know if it does. If we forget and it changes under you, please email support@dataverse.org about it.

File: server/lib/dataverse/provider.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    item = data['items'][0]
+    files = [{
+        'filename': item['name'],
+        'mimeType': item['file_content_type'],
+        'filesize': item['size_in_bytes'],
+        'id': item['file_id'],
+        'doi': 'fixme'
+    }]
+    title = item['name']
+    title_search = _QUOTES_REGEX.search(item['dataset_citation'])
```
Comment (pdurbin) [inline server/lib/dataverse/provider.py]: Please feel free to open an issue at https://github.com/IQSS/dataverse/issues about exposing the title of a dataset separately from the citation in the Dataverse Search API.

Comment (Xarthisius) [reply server/lib/dataverse/provider.py]: Done: https://github.com/IQSS/dataverse/issues/5339

File: server/lib/dataverse/provider.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        title, files, _ = _query_dataverse(search_url)
+        return title, files, doi
+
+    @staticmethod
+    def _parse_access_url(url):
+        """Extract title, file, doi from Dataverse resource.
+
+        Handles: {siteURL}/api/access/datafile/{fileId}
+        """
+        fileId = os.path.basename(url.path)
```
[the remark this answers is not in the document]
Comment (Xarthisius) [reply server/lib/dataverse/provider.py]: The former is now handled. The latter returns a zipfile, which I'd say is out of scope for this PR

File: server/lib/dataverse/provider.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    data = json.loads(resp.decode('utf-8'))['data']
+    if data['count_in_response'] != 1:
+        raise ValueError
+    item = data['items'][0]
+    files = [{
+        'filename': item['name'],
+        'mimeType': item['file_content_type'],
+        'filesize': item['size_in_bytes'],
+        'id': item['file_id'],
+        'doi': 'fixme'
```
[the remark this answers is not in the document]
Comment (Xarthisius) [reply server/lib/dataverse/provider.py]: Thanks for catching that. I wasn't actually setting the `identifier` attribute, that's why it went unnoticed. For now I'm gonna use parent DOI, but I left note pointing to https://github.com/IQSS/dataverse/issues/5339#issuecomment-442168506

**Pull request merged**